Generalized Review System - Milestone 1.1 (Part 3): Suggestion services

nithusha21 opened the pull request:
## Explanation
This PR implements suggestion services. Some parts of this PR (The thread_id generation particularly) will be edited to be simpler once the other PR on fixing the thread_id field goes in. 

PTAL @anmolshkl and @seanlip. Thanks!

## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes.
- [ ] The PR explanation includes the words "Fixes #bugnum: ...".
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

nithusha21 committed:
added suggestion services

nithusha21 committed:
fixed linting

nithusha21 committed:
minor fixes

nithusha21 committed:
minor fixes

seanlip commented on core/domain/feedback_services.py:
```diff
@@ -90,8 +90,10 @@ def create_thread(
-    _create_models_for_thread_and_first_message(
+    thread_id = _create_models_for_thread_and_first_message(
         exploration_id, state_name, original_author_id, subject, text, False)
+    return feedback_models.FeedbackThreadModel.generate_full_thread_id(
```
I thought we were getting rid of all the thread ID / full thread ID stuff?

nithusha21 replied:
Yes @seanlip. I'll edit all these locations after the other PR goes in.

seanlip replied:
Ah, ok. There should be a comment here then, like all the other places. Without a comment there is an implicit assumption that the code is final.

nithusha21 replied:
Oh ok. My bad. Will comment above all locations I plan to edit after that PR goes in. Thanks!

nithusha21 committed:
added comments above lines that will be edited

nithusha21 committed:
removed unnecessary print

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+def create_suggestion(
+        suggestion_type, target_type, target_id, target_version_at_submission,
+        author_id, change_cmd, score_category, description,
+        assigned_reviewer_id='', final_reviewer_id=''):
```
Try and get rid of the default values. Pass in None instead (if relevant); don't use empty strings as a replacement for None.

nithusha21 replied:
Done

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+        suggestion_type: str. The type of the suggestion.
+        target_type: str. The target entity being edited.
+
+        (The above 2 parameters Should be one of the constants defined in
```
Should --> should

nithusha21 replied:
Done

anmolshkl commented on core/domain/feedback_services.py:
```diff
@@ -90,9 +90,13 @@ def create_thread(
+    thread_id = _create_models_for_thread_and_first_message(
         exploration_id, state_name, original_author_id, subject, text, False)
 
+    # Will be changed after the feedback services PR goes in.
```
Sorry, did you mean when #4971 goes in?

nithusha21 replied:
Yes

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+        target_version_at_submission: int. The version number of the target
+            entity at the time of creation of the suggestion.
+        author_id: str. The ID of the user who submitted the suggestion.
+        change_cmd: dict. The actual content of the suggestion.
```
Maybe `content` --> `details` since "content" has a specific meaning in Oppia (it's used for the non-interactive part of a state).

nithusha21 replied:
Done

anmolshkl commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Commands that can be used on suggestions."""
```
This docstring doesn't meaningfully describe the entire `suggestion_services`

anmolshkl replied:
@nithusha21 ping!

nithusha21 replied:
Done

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+            entity at the time of creation of the suggestion.
+        author_id: str. The ID of the user who submitted the suggestion.
+        change_cmd: dict. The actual content of the suggestion.
+        score_category: str. The category to score the suggestor on.
```
nit: I think it's "suggester"?

nithusha21 replied:
Changed to author.

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+        author_id: str. The ID of the user who submitted the suggestion.
+        change_cmd: dict. The actual content of the suggestion.
+        score_category: str. The category to score the suggestor on.
+        description: str. The description of the changes provided by the author.
```
Is "author" the same thing as "suggester"? If so, use the same term to refer to both.

nithusha21 replied:
Done

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+        assigned_reviewer_id: str(optional). The ID of the user assigned to
+            review the suggestion.
+        final_reviewer_id: str(optional). The ID of the reviewer who has
+        accepted/rejected the suggestion.
```
Indent by 4.

nithusha21 replied:
Done

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+        final_reviewer_id: str(optional). The ID of the reviewer who has
+        accepted/rejected the suggestion.
+    """
+    if target_type == suggestion_models.TARGET_TYPE_EXPLORATION:
```
Are you only special-casing this to explorations?

nithusha21 replied:
Till we delink feedback threads from explorations. Will remove this if after that migration is done (As part of milestone 2). Should I add a TODO comment here?

seanlip replied:
Yes please. State when it'll be done, too, and what it's blocking on.

nithusha21 replied:
Added TODO

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+        thread_id = suggestion_models.TARGET_TYPE_EXPLORATION + '.' + thread_id
+    suggestion_models.SuggestionModel.create(
+        suggestion_type, target_type, target_id, target_version_at_submission,
+        suggestion_models.STATUS_IN_REVIEW, author_id, assigned_reviewer_id,
```
Is it actually "in review" if there is no assigned reviewer?

nithusha21 replied:
Hmmm. I guess not. But till we have the automatic assignment of reviewers (part of milestone 2), I think we need to keep the assigned_reviewer field empty on creation. We can probably add a dropdown for owners/editors to assign one among owners/editors for review. But this will also happen after creation. So for the moment, there may be no assigned reviewer though status is in_review.

seanlip replied:
OK. Is it worth having an additional status type, then (like "received")?

nithusha21 replied:
Hmmm. Makes sense, I'll add a new status to distinguish between stuff in review and stuff that's not.

anmolshkl replied:
Sorry, what would "received" indicate?

nithusha21 replied:
Received would indicate suggestion has been created but no reviewer has been assigned to the suggestion.

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+        suggestion_id: str. The ID of the suggestion.
+
+    Returns:
+        Suggestion, or None if no suggestion is found.
```
Suggestion|None. The corresponding suggestion, or None if no suggestion is found.

(Basically, start with type info, then give the description.)

nithusha21 replied:
Done

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+    """Gets a list of suggestions by the given author.
+
+    Args:
+        author_id: str. The author of the suggestions.
```
Here and below: in docstring descriptions, differentiate clearly between the user ID and the user themselves.

nithusha21 replied:
Done

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+        status: str. The status of the suggestion.
+
+    Returns:
+        list(Suggestions): a list of suggestions with the given status.
```
list(Suggestion). A list ...

(Ditto above/below, and also make sure to format docstrings as typeinfo, followed by a full stop, followed by the description.)

nithusha21 replied:
Done

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+
+
+def update_suggestion(suggestion):
+    """Updates the given sugesstion. The properties that can be edited are the
```
"suggestion" is misspelled.

Also, the design here seems a bit weird. I think update_suggestion should be a private function that takes a Suggestion domain object that's already been prepared and just copies the fields blindly to the model. You can then have public functions that correspond to specific actions (like update_reviewer or mark_review_completed, etc.) and that call _update_suggestion in the background.

nithusha21 replied:
I'll make this change to the design. Thanks!

nithusha21 replied:
Actually I have one question here. Wouldn't this increase the number of write operations to the DB? If for each field update we call the _update_suggestion (which writes into the db), that would not be optimal right?

seanlip replied:
I think it depends. It doesn't make sense to do this for multiple changes to the content, yep. But why would you call stuff like update_reviewer or mark_review_completed at the same time?

Basically, figure out what the "high-level actions" are -- mark review completed, start a suggestion, etc. and have a public function for each of them. That function does all the preparations to the domain object -- it doesn't need to limit its changes to just one field. Then it calls _update_suggestion to commit it.

nithusha21 replied:
Ok this makes sense. I'll do this.

I was initially thinking to have one update_status and another update_reviewer and so on. Each of which calls _update_suggestion and then the accept_suggestion calls these functions. But this would have been inefficient. I'll go with your suggestion of having a function for each action.

nithusha21 replied:
Done and made this a protected function.

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+        if state not in states:
+            suggestion.status = suggestion_models.STATUS_INVALID
+            suggestion.final_reviewer_id = reviewer_id
+            suggestion.assigned_reviewer_id = ''
```
Do not use '' to represent None. Ditto elsewhere.

nithusha21 replied:
Done

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+    Returns:
+        bool. Whether the suggestion has been handled or not.
+    """
+    return not suggestion.status == suggestion_models.STATUS_IN_REVIEW
```
Use != rather than "not =="

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+            thread_id = get_thread_id_from_suggestion_id(
+                suggestion.suggestion_id)
+            feedback_services.create_message(
+                suggestion.target_id, thread_id, reviewer_id,
```
It's a bit weird that this message comes from the reviewer_id, since the reviewer didn't write it. Perhaps it should be from a SuggestionValidationBot or something?

nithusha21 replied:
That seems better. I used the pattern followed in the existing framework, but yeah I would go with a bot message too. I'll try to add that.

nithusha21 replied:
Any pointers as to how to add a bot? I tried looking around for other implementations but didn't find any. One idea to have some location in the admin panel where he can click and create a user. But not sure if this is ok. Or on startup, a user (the bot) needs to be created if it doesn't exist (But not sure which file to edit in this case). 

Please let me know what is suitable for this scenario. Thanks!

seanlip replied:
Sorry, I wasn't talking about adding a bot. I was just talking about using a username that made it clear it was a system-generated thing. Grep the codebase for 'OppiaMigrationBot', for example.

nithusha21 replied:
Thanks! I'll add something similar here 👍

nithusha21 replied:
Oh missed this. Will add it soon.

seanlip reviewed (dismissed):
Hi @nithusha21, I've only taken a rough surface-level pass at this, but have made a few comments.

Also, I have one general question: shouldn't we be deleting some code as well in this PR (i.e. the stuff corresponding to the current way suggestions are handled)? Otherwise we just end up with duplication.

anmolshkl commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+        author_id: str. The author of the suggestions.
+
+    Returns:
+        list(Suggestions): a list of suggestions by the given author.
```
nit: A list ...

seanlip replied:
@nithusha21 just a note, it should also be a period instead of a colon. Ditto below.

nithusha21 replied:
Done

anmolshkl commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+        reviewer_id: str. The reviewer of the suggestion.
+
+    Returns:
+        list(Suggestions): a list of suggestions reviewed by the given user.
```
nit: A list ...

nithusha21 replied:
Done

anmolshkl commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+                suggestion.
+
+    Returns:
+        list(Suggestions): a list of suggestions assigned to the given user
```
nit: A list ...

nithusha21 replied:
Done

anmolshkl commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+        status: str. The status of the suggestion.
+
+    Returns:
+        list(Suggestions): a list of suggestions with the given status.
```
nit: A list ...

nithusha21 replied:
Done

anmolshkl commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+        suggestion_type: str. The sub type of the suggestion.
+
+    Returns:
+        list(Suggestions): a list of suggestions of the given type.
```
nit: A list ...

nithusha21 replied:
Done

anmolshkl commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+        target_id: str. The ID of the target entity the suggestion is linked to.
+
+    Returns:
+        list(Suggestions): a list of suggestions linked to the entity.
```
nit: A list ...

nithusha21 replied:
Done

anmolshkl commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+        suggestion_model.final_reviewer_id = suggestion.final_reviewer_id
+        suggestion_model.assigned_reviewer_id = suggestion.assigned_reviewer_id
+
+        suggestion_model.change_cmd = suggestion.change_cmd
```
In which scenario would we able to update the `change_cmd`?

nithusha21 replied:
The change_cmd has an 'old_value' field (for some types of suggestions) which we update to the content just before accepting the suggested value.

Also to allow changes to the suggestion after review, we need to allow the author to edit the suggestion (which is stored in the change_cmd). So then also the change_cmd will be updated.

anmolshkl commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+    if suggestion_model.status == suggestion_models.STATUS_IN_REVIEW:
+        suggestion_model.final_reviewer_id = suggestion.final_reviewer_id
+        suggestion_model.assigned_reviewer_id = suggestion.assigned_reviewer_id
+
```
blank line?

nithusha21 replied:
Removed

anmolshkl commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+        suggestion_model.change_cmd = suggestion.change_cmd
+
+    suggestion_model.status = suggestion.status
+
```
It is OK if you skip this blank line.

nithusha21 replied:
Done

anmolshkl commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+        bool. The validity of the suggestion.
+    """
+
+    if (
```
I fear that we might end up with a lot if....else statements if we do this for every suggestion type. Ideally, we could have kept the validation method specific to every suggestion if every suggestion had its own class.

nithusha21 replied:
Discussed this over the meeting. Decided to use sub-classes. #4989 has been created.

nithusha21 replied:
Done

anmolshkl commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+
+
+def is_suggestion_handled(suggestion):
+    """Checks if the suggestion has been handled.
```
Be more precise. What does "handling" mean?

nithusha21 replied:
Done

anmolshkl reviewed (changes requested):
@nithusha21 took a partial look. Will review the other half by tonight!

nithusha21 committed:
made sub class for types of suggestion and added validate()

nithusha21 committed:
added get_model_corresponding_to_suggestion function and tests, linting fixes

nithusha21 committed:
added docstring

nithusha21 committed:
simplified test

nithusha21 commented:
I thought about the possible things to abstract from this PR. Going forward with the subclassing for various types of suggestions, I think the following things can be done.
- make validators separate for each type of suggestion and write them in the subclass
- add a function in the subclass to accept the suggestion, that just "applies" the change blindly. 
- In the services accept_suggestion will not have the if..else statements and instead would validate, and then accept the suggestion. i.e. it would be general.

Does this answer your question about stuff that can end up being repetitive @seanlip?

seanlip commented:
Hi @nithusha21, which question of mine are you referring to in your latest general comment? (I'm not sure.) Could you link to it? Or are you referring to @anmolshkl's comment [here](https://github.com/oppia/oppia/pull/4982#discussion_r191069473)? In any case, the changes you propose sound sensible to me and I'm fine with them.

That said, if you're referring to my earlier general question about "why has nothing been deleted yet", then I think that question still stands and isn't resolved by your latest comment.

nithusha21 commented:
Yes I was referring to your general comment https://github.com/oppia/oppia/pull/4982#pullrequestreview-123576100. I haven't made any of these changes yet. But was this the intended line of action? (Wasn't sure from your comment what stuff needed to be deleted).

seanlip commented:
I don't know what the intended line of action is. I guess my question is more along the lines of: when are you planning to move over from the current suggestions framework to the one in this PR?

I.e., generally speaking, we already have some code for dealing with suggestions. At which point do we stop using that code and start using the code here?

nithusha21 commented:
Oh ok. Here is my plan. We make the services and controllers for the new framework. Then we wire up the frontend to the new framework. In the release where the new framework goes into production, we will run a one-off job to convert all old suggestions to new suggestions. Once this migration is successful, we delete the old code for the old suggestions framework.

seanlip commented:
Hm. I think you'll want a flag along the lines of ENABLE_NEW_SUGGESTIONS_FRAMEWORK that can be flipped by the release coordinator, then. Pushing a new release shouldn't in itself be construed as turning on the flag, that's hard to control. And besides the release coordinator needs to run the migration job before turning on the new system anyway.

Can you prepare a release instructions doc for @tjiang11, who is coordinating the next release? Please check in with @pranavsid98 as well on how to do this; he's quite familiar with this sort of thing and should be able to give you guidance. We need to make sure there's a clear exact sequence of steps to be followed so that we don't lose or corrupt any data.

nithusha21 commented:
Oh ok. I'll try adding the flag as you suggested. Will talk to Pranav too. Thanks!

nithusha21 committed:
added accept and from_dict functions

nithusha21 committed:
added not implemented checks

nithusha21 committed:
made more comprehensive validators

nithusha21 committed:
fixed linting

nithusha21 committed:
review changes

nithusha21 committed:
minor fix

nithusha21 committed:
linting fixes

nithusha21 committed:
linting fixes

nithusha21 committed:
made review changes

nithusha21 committed:
minor fix

nithusha21 committed:
minor fix to validators

nithusha21 committed:
linting fix

nithusha21 committed:
Merge remote-tracking branch 'origin/develop' into suggestion-services

nithusha21 committed:
Merge remote-tracking branch 'origin/develop' into sub-classing-suggestion-domain

nithusha21 committed:
Merge branch 'sub-classing-suggestion-domain' into suggestion-services

nithusha21 committed:
review changes

anmolshkl commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+(feedback_models, suggestion_models) = models.Registry.import_models([
+    models.NAMES.feedback, models.NAMES.suggestion])
+
+DEFAULT_SUGGESTION_THREAD_SUBJECT = 'Suggestion from a learner'
```
Would "learner" be valid in all context?

nithusha21 replied:
Hmmm. Actually no. I'll have to think of a better default subject

nithusha21 replied:
Changed learner to user. If you have any better ideas please suggest. Thanks!

nithusha21 committed:
minor fixes

anmolshkl commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+
+
+def get_suggestion_from_model(suggestion_model):
+    """Converts the given SuggestionModel to a Suggestion object
```
*Suggestion domain object.

nithusha21 replied:
Done

anmolshkl commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+    """Gets a list of suggestions with the given sub_type.
+
+    Args:
+        suggestion_type: str. The sub type of the suggestion.
```
Are we still calling it sub-type?

nithusha21 replied:
Edited

anmolshkl commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+
+
+def get_suggestion_by_type(suggestion_type):
+    """Gets a list of suggestions with the given sub_type.
```
Is it type or sub type?

nithusha21 replied:
Oops, its type. Will edit in both places

nithusha21 replied:
Done

nithusha21 committed:
fixed linting

anmolshkl commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,367 @@
+        str. The thread ID linked to the suggestion.
+    """
+
+    # Temporarily. Once the feedback migration is complete, this function will
```
Please add a TODO.

nithusha21 replied:
Done

anmolshkl commented on core/domain/suggestion_registry.py:
```diff
@@ -0,0 +1,329 @@
+            review the suggestion.
+        final_reviewer_id: str. The ID of the reviewer who has accepted/rejected
+            the suggestion.
+        change_cmd: dict. The actual content of the suggestion.
```
As Sean had mentioned earlier, avoid using "content"

nithusha21 replied:
Done

anmolshkl commented on core/domain/suggestion_registry.py:
```diff
@@ -0,0 +1,329 @@
+            'Subclasses of BaseSuggestion should implement from_dict.')
+
+    def get_score_type(self):
+        """Returns the first part of the score category.
```
I think it would be worth to describe what the first part actually means or could mean.

nithusha21 replied:
Done

anmolshkl commented on core/domain/suggestion_registry.py:
```diff
@@ -0,0 +1,329 @@
+            suggestion_models.SCORE_CATEGORY_DELIMITER)[0]
+
+    def get_score_sub_type(self):
+        """Returns the second part of the score category.
```
Ditto for the second part.

nithusha21 replied:
Done

anmolshkl commented on core/domain/suggestion_registry.py:
```diff
@@ -0,0 +1,329 @@
+                    suggestion_models.SCORE_CATEGORY_DELIMITER,
+                    self.score_category))
+
+
```
Extra blank line?

nithusha21 replied:
Done

anmolshkl commented on core/domain/suggestion_registry.py:
```diff
@@ -0,0 +1,329 @@
+        exploration = exp_services.get_exploration_by_id(self.target_id)
+        old_content = exploration.states[self.change_cmd['state_name']].content
+
+        self.change_cmd['old_content'] = old_content
```
Where are we thinking of putting the validations to check that old_content has not changed after making the suggestion?

nithusha21 replied:
For this type of suggestion we don't need the check. What this function does is it stores the value of the content just before accepting as the old_content (so that if the change is reverted it goes back to the correct content). 

When we create the translations sub class, in its validate() we will have the check for old content being same as current content. The validate() is called just before accepting.

anmolshkl replied:
Doesn't that apply to exploration content as well? If the old_content has changed significantly, I don't think that suggestion will be valid. But we can leave that for the reviewers to decide.

nithusha21 replied:
Yeah. The reviewer can reject if it is out of context since when the suggestion was made. This is not something (I think) we can detect in code. If the content is different but the state is still present, theoretically, the suggestion can still be accepted.

anmolshkl commented on core/domain/suggestion_registry.py:
```diff
@@ -0,0 +1,329 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Registry for Oppia suggestions. Contains a BaseSuggestion class and
```
Though you can call it a registry but doesn't strictly looks/works like a registry used in Oppia (for e.g. the interaction registry).

nithusha21 replied:
Discussed it on chat. Decided we'll go ahead with this for now.

anmolshkl commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,384 @@
+            accepted/rejected the suggestion.
+    """
+
+    # TODO: Remove the check for target type once the feedback threads are
```
TODO(nithusha21)

nithusha21 replied:
Done

anmolshkl commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,384 @@
+
+
+def get_suggestion_from_model(suggestion_model):
+    """Converts the given SuggestionModel to a Suggestion object
```
..Suggestion domain object.

nithusha21 replied:
Done

nithusha21 committed:
made review changes

nithusha21 committed:
made review changes

nithusha21 committed:
Merge remote-tracking branch 'origin/develop' into suggestion-services

nithusha21 committed:
more review changes

nithusha21 committed:
more review changes

nithusha21 commented:
@seanlip and @anmolshkl I have merged the changes of #4989 into this branch. Also I have addressed the review comments (except one about adding a SuggestionValidationBot, i'll add this soon)

nithusha21 committed:
Fix backend tests -- Merge remote-tracking branch 'origin/develop' into suggestion-services

seanlip commented on core/domain/suggestion_registry.py:
```diff
@@ -0,0 +1,332 @@
+        self.change_cmd['old_content'] = old_content
+
+        self.change_cmd['new_content']['audio_translations'] = (
+            old_content['audio_translations'])
```
/cc @DubeySandeep -- this will need updating if @nithusha21's PR goes in before yours.

seanlip commented on core/domain/suggestion_registry.py:
```diff
@@ -0,0 +1,332 @@
+
+    @classmethod
+    def from_dict(cls):
+        """Return a Suggestion object of type from a dict."""
```
No idea what this means: "of type from a dict"?

seanlip commented on core/domain/suggestion_registry.py:
```diff
@@ -0,0 +1,332 @@
+                    'Expected final_reviewer_id to be a string, received %s' %
+                    type(self.final_reviewer_id))
+
+        if not isinstance(self.change_cmd, dict):
```
This should be a Change object actually (TopicChange, ExplorationChange, etc.). Check with @pranavsid98 and/or @aks681 for details, and see Akshay's recent message on oppia-dev@ too.

nithusha21 replied:
Done

seanlip commented on core/domain/suggestion_registry.py:
```diff
@@ -0,0 +1,332 @@
+            raise utils.ValidationError(
+                'Expected change_cmd to contain a state_name key')
+
+        states = exp_services.get_exploration_by_id(self.target_id).states
```
Drop this check, otherwise if the state name changes and you make some subsequent change to the suggestion (e.g. change the reviewer ID) you won't be able to save it. Instead, do this check at the point of suggestion acceptance only (maybe have a separate method for that). Also make sure (going forward) to figure out how to deal with obsolete state names somehow in the suggestion UI.

nithusha21 replied:
If the state name changes, and the state the suggestion was made to is invalid, shouldn't we invalidate it? That was my idea behind this check. Later we can try to convert invalid suggestions to valid ones by letting the author choose which state he meant to suggest to(post rename of state). But in case the state was deleted itself, the suggestion is truly invalid.

So if a suggestion is invalid, shouldn't we set status to invalid ASAP to prevent reviews and changes and then realising that the suggestion can't be accepted?

seanlip replied:
But the trigger for the state name becoming invalid is the exploration commit, not a change to the suggestion. The effect of what you're doing here is just to prevent the suggestion from being edited further. (Right?)

nithusha21 replied:
That is true. If validate() fails with an exception, then the suggestion cannot be accepted. I think it is logical to also prevent further edits to a suggestion if it is invalid. However, another possibility is in future, if we design a mechanism to make invalid suggestions valid (I feel this is similar to resolving merge conflicts!) then we don't want to prevent further edits and rather just validate just before accepting and if invalid, allow the author to edit in order to fix these issues.

seanlip replied:
OK, but for now, maybe just have a message in frontend saying "This suggestion cannot be accepted because ..." and gray out the accept button, leaving only the reject button?

nithusha21 replied:
Moved this to pre_accept_validate()

seanlip commented on core/domain/suggestion_registry.py:
```diff
@@ -0,0 +1,332 @@
+        """Before accepting the suggestion, modifications need to be done to the
+        change_cmd.
+        """
+        exploration = exp_services.get_exploration_by_id(self.target_id)
```
Better to pass the old_content in as an argument, if possible.

Also don't modify the domain object, do something like get_change_list_for_accepting_suggestion() or similar and let the *_services.py file handle all the operations like calling update_exploration() etc.

nithusha21 replied:
Actually I think passing it separately won't work. For every suggestion type the old content is obtained differently. So I think obtaining it within this function is a better idea for generalizing. What's your opinion on this?

Also I did make a get_change_list_for_accepting_suggestion function which I call just before accepting. Another thing I am doubtful about is letting the services file handle update_exploration. This again relates to generalisation. It would be better to have it in the accept function in the domain object itself as it is type specific.

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,382 @@
+        change_cmd: dict. The details of the suggestion.
+        score_category: str. The category to score the author on.
+        description: str. The description of the changes provided by the author.
+        assigned_reviewer_id: str(optional). The ID of the user assigned to
```
Typeinfo should be "str or None." Ditto below.

(I mean, it's clearly not optional, there's no default arg. And there probably shouldn't be.)

nithusha21 replied:
Done

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,382 @@
+            DEFAULT_SUGGESTION_THREAD_SUBJECT)
+        # This line will be removed after the feedback thread migration is
+        # complete and the IDs for both models match.
+        thread_id = suggestion_models.TARGET_TYPE_EXPLORATION + '.' + thread_id
```
else raise exception

nithusha21 replied:
Done

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,382 @@
+            target_version_at_submission, suggestion_models.STATUS_IN_REVIEW,
+            author_id, assigned_reviewer_id, final_reviewer_id, change_cmd,
+            score_category, thread_id)
+    else:
```
Find a way not to branch here. So much of this is the same between both branches. Define a variable for the status (`status = ... if ... else ...`) and use that.

In general, repeat yourself as little as possible. This is a very important principle in software engineering.

nithusha21 replied:
Done. Thanks!

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,382 @@
+    Args:
+        suggestion: Suggestion. The domain object of the suggestion to be
+            updated.
+        status: str. The status of the suggestion post review.
```
Perhaps list (and/or validate) the allowed choices

nithusha21 replied:
Done

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,382 @@
+    """Updates the given suggestion.
+
+    Args:
+        suggestion: Suggestion. The domain object of the suggestion to be
```
The suggestion to be updated.

(When talking about nouns the default assumption is that they're domain objects.)

nithusha21 replied:
Done

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,382 @@
+    return suggestion_id[suggestion_id.find('.') + 1:]
+
+
+def is_suggestion_handled(suggestion):
```
Seems more like suggestion.is_handled? (define as a `@property`)

Especially since it is primarily dependent on a property of the suggestion object.

nithusha21 replied:
Done

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,382 @@
+        raise Exception('Commit message cannot be empty.')
+
+    try:
+        # If an exception is raised, the suggestion status will be set as
```
That's a bit weird. I think it needs to be more granular than this, and perhaps a separate message defined for this check. What exactly is the concern here?

If it's e.g. a state name changing then have a separate method devoted to actually checking that and other stuff that you want to check at this juncture. Otherwise you end up shadowing errors that should really be caught and fixed.

nithusha21 replied:
With reference to the [above](https://github.com/oppia/oppia/pull/4982#discussion_r192581023) comment, I think a general validation of the suggestion needs to be done before accepting to make sure that everything is fine. And yes I think this process does not actually tell you what the error is and instead just says the suggestion is valid. 

For the moment I don't have a design for getting suggestions out of an invalid status. But I wanted to mark status as STATUS_INVALID and this status change should be brought about if an exception is raised in the validate(). This was the best approach I had in mind, any suggestions for a better approach?

seanlip replied:
We could have a cron job that iterates through suggestions and invalidates them periodically, but that feels like overkill. I think the question really is _when_ to mark a suggestion invalid and I think it needs to be initiated by someone -- so, if not due to a commit, it must be initiated by whoever is triaging the suggestion.

So I think for now it's reasonable to let the triager actually reject the suggestion manually along the lines of my comment above. Validity checks prior to regular saves of the suggestion object should be reserved for actual structural integrity (like the suggestion is formatted correctly and makes sense as a suggestion object) but they do not need to check referential integrity. The latter is checked on the fly in the frontend and at the time of acceptance of a suggestion.

nithusha21 replied:
Ok I see your point. So basically while updating suggestions we need to validate that parameter X is actually a string and so on. But for the second point, don't we actually need a validation in the backend too? To check that the suggestion references valid properties/values to be edited? Yeah this probably should go into a separate validator like a pre_accept_validator or similar. 

If we want to invalidate/validate when commits are made. An important question arises -- If a state is added with name same as a previously deleted state, should all suggestions to the previously deleted state become valid again? I would go with your suggested approach of greying out the accept button with a message saying "the suggestion can't be accepted because of XYZ reason please ask the author to make a new suggestion to a valid state."

seanlip replied:
Yes, we do need backend validation for referential integrity but this only really matters on suggestion acceptance. So we can do it then.

I like the message you suggest, except that you might not need the "please ask the author" part. You could prepopulate the rejection message to explain what happened though. Remember that the suggester cannot see the state and doesn't even know the state name.

Good question about a state being added with the same name as a previously deleted state. It's hard to handle this, short of using state IDs (cf. @prasanna08's work), but it doesn't really matter too much since the creator is looking at the old/new content anyway. So I think it's fine if the invalid comments become valid again and subsequently get rejected.

seanlip replied:
Well, actually, "fine" is a strong word. It's suboptimal, but I don't see a way around it except to listen to commits, which seems expensive.

nithusha21 replied:
So shall we get rid of STATUS_INVALID concept altogether? and have the reject message indicate that it was invalid?

seanlip replied:
I don't know what you _should_ do (probably want to ask @anmolshkl who has more context), but I have no immediate objection to what you said.

anmolshkl replied:
I agree with @seanlip. It is better to leave the validation of structural integrity till the end i.e. when a suggestion is accepted/rejected. Also, I don't think the renaming of states could be a major problem since the decision would be made by the reviewer as to whether the suggestion is valid or not (assuming we will be pulling the latest contents of the state based on its name). If the state name is not found, you could indicate that it is an invalid suggestion or else just show the state contents to the reviewer, in which case, he can decide if the old content and the new content make sense or not.

nithusha21 replied:
Yeah. Makes sense. I'll separate this check out from validate and also in the frontend, we'll validate the suggestion when displayed to a reviewer and grey out the accept button if not valid. This removes the necessity of the invalid flag to be stored (We can determine it during execution). So I'll remove that too.

nithusha21 replied:
Removed the invalid status, split the validation into separate functions for structural and referential validity. Also we'll have a check in the frontend as mentioned.

seanlip commented on core/domain/suggestion_registry.py:
```diff
@@ -0,0 +1,332 @@
+        self.change_cmd['new_content']['audio_translations'] = (
+            old_content['audio_translations'])
+
+    def accept(self, commit_message):
```
Is this tested anywhere?

nithusha21 replied:
This function is actually a tricky one to unit test. I thought about it for a while and I'm not sure how to test this. I decided that when we write a final integration test for the framework, we can test the behavior effectively. 

So to unit test the issue is we have to create an exploration, add a state, and create a valid suggestion, and then accept suggestion. Once accepted, validate new content is the same as the suggested content -- This feels more like an integration test rather than a unit test. Also the accept_suggestion (service) which calls this method is tested by mocking update_exploration.

What do you suggest should be done here?

seanlip replied:
Yeah, I would expect an integration test (without mocking update_exploration). Is there a reason we can't do that now?

nithusha21 replied:
Nothing blocking this. I thought we'll add the integration test at the controller level (by sending requests to the appropriate URLs). But I think it makes sense to have an integration here too. Will add one.

nithusha21 replied:
Still working on this. Will push this soon.

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,382 @@
+    feedback_services.create_message(
+        get_thread_id_from_suggestion_id(suggestion.suggestion_id), reviewer_id,
+        feedback_models.STATUS_CHOICES_FIXED, None,
+        'Accepted by %s' % reviewer_id)
```
Don't we already have a default thing show up when the status of a thread changes? Maybe the message should be an optional addition by the reviewer?

nithusha21 replied:
Yeah. The message shows up as User_name <gap> Status changed to 'Fixed' <next line> (message). So I think by default the message can be blank and if the reviewer wants to add something else then we can pass it here. Also with this view I think the bot isn't necessary because the way it shows up makes sense (The user did actually change the status and optionally give the message).

nithusha21 replied:
Passes review_message argument. If None, only the default message will be shown. If something is passed, that will be shown in the thread.

seanlip commented on core/domain/suggestion_services.py:
```diff
@@ -0,0 +1,382 @@
+    feedback_services.create_message(
+        get_thread_id_from_suggestion_id(suggestion.suggestion_id), reviewer_id,
+        feedback_models.STATUS_CHOICES_IGNORED, None,
+        'Rejected by %s' % reviewer_id)
```
Here, I think the message should definitely be a (mandatory) addition by the reviewer to indicate what the problem is.

nithusha21 replied:
Added a new parameter as mentioned above. Raised exception if the parameter is None.

seanlip reviewed (commented):
Only took a brief look. @anmolshkl could you take a pass please?

anmolshkl commented on core/domain/suggestion_registry.py:
```diff
@@ -0,0 +1,332 @@
+            status, author_id, assigned_reviewer_id, final_reviewer_id,
+            change_cmd, score_category):
+        """Initializes a Suggestion object of type
+        SUGGESTION_TYPE_EDIT_STATE_CONTENT.
```
The object is of type SuggestionEditStateContent. Perhaps, it is better to say "Initializes an object of type SuggestionEditStateContent corresponding to the SUGGESTION_TYPE_EDIT_STATE_CONTENT choice" or something along those lines.

nithusha21 replied:
Done

anmolshkl commented on core/domain/suggestion_registry.py:
```diff
@@ -0,0 +1,332 @@
+        self.score_category = score_category
+
+    def validate(self):
+        """Validates a suggestion object of type
```
Perhaps, just say validates the SuggestionEditStateContent object.

nithusha21 replied:
Done

anmolshkl commented on core/domain/suggestion_registry.py:
```diff
@@ -0,0 +1,332 @@
+
+    @classmethod
+    def from_dict(cls, suggestion_dict):
+        """Return a Suggestion object of type from a dict.
```
of type from a dict? What does it mean?

anmolshkl commented on core/domain/suggestion_registry.py:
```diff
@@ -0,0 +1,332 @@
+
+    @classmethod
+    def from_dict(cls, suggestion_dict):
+        """Return a Suggestion object of type from a dict.
```
of type from a dict? What does it mean?

nithusha21 replied:
Edited

anmolshkl commented on core/domain/suggestion_registry.py:
```diff
@@ -0,0 +1,332 @@
+            suggestion_dict: dict. The dict representation of the suggestion.
+
+        Returns:
+            BaseSuggestion. The corresponding Suggestion domain object.
```
Can be called a BaseSuggestion but it is better to refer to the specific type i.e. SuggestionEditStateContent

nithusha21 replied:
Done

anmolshkl commented on core/domain/suggestion_registry.py:
```diff
@@ -0,0 +1,332 @@
+        return suggestion
+
+
+SUGGESTION_TYPES_TO_DOMAIN_CLASSES = {
```
SUGGESTION_TYPES_TO_DOMAIN_CLASS_MAPPING

seanlip replied:
Just my 2c, I'm also fine with the existing name (since it's starting to get a bit long)

nithusha21 replied:
I didn't edit it. Please let me know if it is fine how it is or it needs a change. Thanks!

anmolshkl replied:
I think it is fine to leave that as it is.

anmolshkl reviewed (changes requested):
@nithusha21 left a couple of comments along with the review. PTAL!

nithusha21 committed:
merge develop

nithusha21 committed:
review changes (1)

nithusha21 committed:
linting fixes

nithusha21 committed:
more linting fixes

nithusha21 committed:
minor fix

anmolshkl commented on core/domain/suggestion_registry.py:
```diff
@@ -0,0 +1,348 @@
+            bool. Whether the suggestion has been handled or not.
+        """
+        return (self.status not in [suggestion_models.STATUS_IN_REVIEW,
+                                    suggestion_models.STATUS_RECEIVED])
```
What does STATUS_RECEIVED represent? Does it refer to the scenario where suggestion has been received, but no reviewer has been assigned and hence, it can't be said to be in review?

nithusha21 replied:
Yes

anmolshkl commented on core/domain/suggestion_registry.py:
```diff
@@ -0,0 +1,348 @@
+        states = exp_services.get_exploration_by_id(self.target_id).states
+        if self.change_cmd.state_name not in states:
+            raise utils.ValidationError(
+                'Expected %s to be a valid state name' %
```
@nithusha21  What if I re-name the state but do not change the content (which I have actually done in the past)? This would mean a suggestion (which is still valid) would be invalidated?

nithusha21 replied:
Yes the suggestion would be invalidated as it is not referentially valid anymore. 

The possible user flow that will happen here is the reviewer rejects this suggestion and then the author creates a new suggestion to the (new, renamed) state and then the reviewer can accept it.

However there is one complication of what message to show the author (as state names are only known to the creators/editors).

nithusha21 committed:
fix backend tests

nithusha21 committed:
added basic integration test

nithusha21 committed:
linting fixes

nithusha21 committed:
more linting fixes

nithusha21 commented:
Added an integration test to test suggestion creation and accepting. Nothing was mocked in the process except the thread_id generation (to prevent the randomness). PTAL. Thanks!

nithusha21 committed:
added test for rejection and accept with message

anmolshkl commented on core/domain/suggestion_registry_test.py:
```diff
@@ -0,0 +1,147 @@
+(suggestion_models,) = models.Registry.import_models([models.NAMES.suggestion])
+
+
+class SuggestionRegistryUnitTests(test_utils.GenericTestBase):
```
Umm I would rather prefer to keep unit tests specific to the domain classes than to the entire registry. Otherwise, if in the future, when you add let us say SuggestionEditStateHint, where will you add the corresponding units tests?

nithusha21 replied:
Split up the test suite for specific classes.

anmolshkl commented on core/domain/suggestion_services_test.py:
```diff
@@ -0,0 +1,602 @@
+        return None
+
+
+    def return_true(self):
```
I see that this method is never used.

nithusha21 replied:
Removed

anmolshkl commented on core/domain/suggestion_services_test.py:
```diff
@@ -0,0 +1,602 @@
+    def return_true(self):
+        return True
+
+    def return_false(self):
```
ditto - method never used

nithusha21 replied:
Removed

anmolshkl reviewed (changes requested):
@nithusha21 almost there! Just a couple of suggestions and changes.

nithusha21 committed:
made review changes

nithusha21 committed:
linting fix

anmolshkl commented:
@nithusha21 LGTM! (y)

anmolshkl reviewed (approved):
@nithusha21 lgtm!

nithusha21 commented:
Thanks @anmolshkl! @seanlip could you also take a look. Thanks!

seanlip commented on core/domain/suggestion_registry.py:
```diff
@@ -0,0 +1,348 @@
+            exploration.states[self.change_cmd.state_name].content.to_dict())
+
+        change_cmd.old_value = old_content
+        change_cmd.new_value['audio_translations'] = (
```
One thought: if the content changes, wouldn't the audio translations need to be marked stale in some cases? Maybe you'll want the reviewer to indicate via a checkbox whether this is a "big" change or not, and act accordingly here.

Could be a separate PR, but make sure to track it.

nithusha21 replied:
Oh true. So the first thought was to pass in an extra parameter to accept_suggestion which says whether or not to mark these translations as needing update. However this parameter would not make sense in some suggestion scenarios which do not involve translations (like say add question to topic). So is it ok to pass in None in those situations (I don't think this affects the generality too much though).

Also yeah. I think I'll push it in a separate PR (as this blocks the next PR for this milestone).

seanlip replied:
It's unclear to me whether this should be considered a modification of the changelist, or whether the changelist should come from the editor frontend, or something else. I lean towards finishing up the e2e experience for now and then adding this in as a feature after that (please file an issue).

seanlip replied:
(Also feel free to merge if you're happy with this :) )

nithusha21 replied:
Yep. Will continue further tasks and complete the full experience then decide what's appropriate here. Will file an issue. 

Merging thanks!

seanlip reviewed (commented):
I just took a quick look but basically defer to @anmolshkl's review. If he says it's OK then it's fine by me :) 

(I'll dismiss my review so it doesn't block.)
